### ISSUE TITLE
Sort.sortObj() is not safe with mixed boxed types

### ISSUE THREAD

**Issue by niloc132** [deephaven/deephaven-core#4570]
### Example groovy 1
```groovy
print(sortObj(1, 0.1))
```
#### Expected
 `[0.1, 1.0]`
#### Actual
```
 java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.math.BigDecimal (java.lang.Integer and java.math.BigDecimal are in module java.base of loader 'bootstrap')
        at java.base/java.math.BigDecimal.compareTo(BigDecimal.java:232)
        at io.deephaven.function.comparators.NullNaNAwareComparator.compare(NullNaNAwareComparator.java:59)
        at io.deephaven.function.comparators.NullNaNAwareComparator.compare(NullNaNAwareComparator.java:17)
        at java.base/java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
        at java.base/java.util.TimSort.sort(TimSort.java:220)
        at java.base/java.util.Arrays.sort(Arrays.java:1441)
        at io.deephaven.function.Sort.sortObj(Sort.java:75)
        at io.deephaven.function.Sort.sortObj(Sort.java:87)
...
```
### Example groovy 2
Also unsafe with int/double:
```groovy
print(sortObj(1, .1d))
```
#### Expected
`[0.1, 1]`

#### Actual
```
java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.Double (java.lang.Integer and java.lang.Double are in module java.base of loader 'bootstrap')
        at java.base/java.lang.Double.compareTo(Double.java:49)
        at io.deephaven.function.comparators.NullNaNAwareComparator.compare(NullNaNAwareComparator.java:59)
        at io.deephaven.function.comparators.NullNaNAwareComparator.compare(NullNaNAwareComparator.java:17)
        at java.base/java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
        at java.base/java.util.TimSort.sort(TimSort.java:220)
        at java.base/java.util.Arrays.sort(Arrays.java:1441)
        at io.deephaven.function.Sort.sortObj(Sort.java:75)
        at io.deephaven.function.Sort.sortObj(Sort.java:87)
...
```

### Example in query language (from python)
```python
from deephaven import empty_table
d = 0.1
t = empty_table(3).update(['A = i', 'Values = sortObj(new Comparable[]{d, A})'])
```
Note that this is a bit more code than it probably should need to be, but reflects a test case represented in DHE, where it is possible to use `Number[]` (the analogous method has special casing for Number arrays).
#### Expected
```
A	Values
0	[0.0,0.1]
1	[0.1,1.0]
2	[0.1,2.0]
```
#### Actual
```
java.lang.ClassCastException: class java.lang.Double cannot be cast to class java.lang.Integer (java.lang.Double and java.lang.Integer are in module java.base of loader 'bootstrap')
        at java.base/java.lang.Integer.compareTo(Integer.java:59)
        at io.deephaven.function.comparators.NullNaNAwareComparator.compare(NullNaNAwareComparator.java:59)
        at io.deephaven.function.comparators.NullNaNAwareComparator.compare(NullNaNAwareComparator.java:17)
        at java.base/java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
        at java.base/java.util.TimSort.sort(TimSort.java:220)
        at java.base/java.util.Arrays.sort(Arrays.java:1441)
        at io.deephaven.function.Sort.sortObj(Sort.java:75)
        at io.deephaven.function.Sort.sortObj(Sort.java:87)
        at io.deephaven.temp.c_904c6956e578c82f18b8b45f1825161c01dc9a04e026062178d417a4f7eb8ebev55_0.Formula.applyFormulaPerItem(Formula.java:153)
```

**Comment by rcaudy** [comment]
This seems kind of like an obvious limitation. What do you want to change, besides maybe a better error message?

**Comment by chipkent** [comment]
I agree with @rcaudy that the code is behaving as expected.  We used to have a more complex operation that handled type conversion for all kinds of numerics and such, but it was deemed to be a crime against humanity, so we went with the current, simplified version.

The signature of `sortObj` is `public static <T extends Comparable<? super T>> T[] sortObj(ObjectVector<T> values)`.  Because of Java's awful generics, it is hard to do better without hard coding in a bunch of special cases within [`NullNaNAwareComparator`](https://github.com/deephaven/deephaven-core/blob/main/engine/function/src/main/java/io/deephaven/function/comparators/NullNaNAwareComparator.java)

If you use `sort` and not `sortObj`, you get automatic conversion as expected.

```python
from deephaven import empty_table
d = 0.1
t = empty_table(3).update(['A = i', 'Values = sort(d, A)'])
```
![image](https://github.com/deephaven/deephaven-core/assets/5250374/8e79975a-a0bf-46ad-8cd8-8e0b1d7008a1)

We could do better with more advanced formula translation in the query engine, but that is a much more difficult problem.

Maybe we could convert any number to `BigDecimal` and compare.
